### PR TITLE
[incremental build] first draft of joint compilation, with java source language

### PR DESCRIPTION
**Pull request by svenefftinge**

### CODE REVIEW

File: plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+					val builder = new JvmDeclaredTypeBuilder(new BinaryClass(topLevel, inMemClassLoader), new ClassFileBytesAccess(), inMemClassLoader)
+					resource.contents += builder.buildType
+				}
+			}
+		], new DefaultProblemFactory())
+		compiler.compile(#[compilationUnit])
+	}
+	
+	protected def CompilerOptions getCompilerOptions() {
+		val jdkVersion = ClassFileConstants.JDK1_7;
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend]: Does this work with Java8 code, e.g. for repeatable annotations?

File: plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	
+	@Inject IEncodingProvider encodingProvider
+	@Inject CompilerPhases compilerPhases
+	@Inject JavaDerivedStateComputer derivedStateComputer
+	
+	CompilationUnit compilationUnit
+	
+	override protected doLoad(InputStream inputStream, Map<?, ?> options) throws IOException {
+		// add a dummy eobject as root
+		val encoding = encodingProvider.getEncoding(getURI)
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend]: The encoding may have been passed as an option to load(..), see XtextResource.setEncodingFromOptions(Map<?, ?>)

File: plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import org.eclipse.jdt.internal.compiler.batch.CompilationUnit
+import org.eclipse.xtext.common.types.JvmTypeReference
+import org.eclipse.xtext.common.types.access.IJavaSchemeUriResolver
+import org.eclipse.xtext.common.types.access.impl.AbstractJvmTypeProvider
+import org.eclipse.xtext.common.types.access.impl.IndexedJvmTypeAccess
+import org.eclipse.xtext.common.types.access.impl.URIHelperConstants
+import org.eclipse.xtext.parser.IEncodingProvider
+import org.eclipse.xtext.resource.CompilerPhases
+
+class JavaResource extends ResourceImpl implements IJavaSchemeUriResolver {
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend]: JvmGenericType checks the resource for ISynchronizable - maybe this resource should implement that interface and synch on getContents, too

File: plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+				return result;
+			}
+		}
+		return EcoreUtil.resolve(proxy, sender)
+	}
+	
+	IndexedJvmTypeAccess _access
+	def IndexedJvmTypeAccess getIndexJvmTypeAccess() {
+		if (_access == null) {
+			val provider = resourceSet.getResourceFactoryRegistry().getProtocolToFactoryMap().get(URIHelperConstants.PROTOCOL)
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend]: we could probably make org.eclipse.xtext.common.types.access.impl.AbstractJvmTypeProvider.getIndexedJvmTypeAccess() public

File: plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend (hunk 1 above)
Comment (svenefftinge) [inline plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend]: The compiler options will need to be configured from outside. They have to match whatever the project uses. This is just a default for the time being.

**Pull request merged**